### PR TITLE
azure-vnet: Implement GET command

### LINKED ISSUE #148: How to build this?
**Is this a request for help?**:
YES
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one):

---

**Which release version?**:
master
---

**Which component (CNI/IPAM/CNM/CNS)**:

---

**Which Operating System (Linux/Windows)**:

Centos 7 / Linux localhost 3.10.0-693.21.1.el7.x86_64 #1 SMP Wed Mar 7 19:03:37 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
---

**Which Orchestrator and version (e.g. Kubernetes, Docker)**

**What happened**:
make all-binaries fails

**What you expected to happen**:

make all-binaries success

**How to reproduce it** (as minimally and precisely as possible):

```
$ git clone https://github.com/Azure/azure-container-networking
$ cd azure-container-networking
$ make all-binaries
```
**Anything else we need to know**:

```
[user@localhost ~]$ git clone https://github.com/Azure/azure-container-networking
Cloning into 'azure-container-networking'...
remote: Counting objects: 2482, done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 2482 (delta 0), reused 2 (delta 0), pack-reused 2476
Receiving objects: 100% (2482/2482), 4.33 MiB | 3.73 MiB/s, done.
Resolving deltas: 100% (1448/1448), done.
```

```
cd azure-container-networking/
```

```
[user@localhost azure-container-networking]$ make all-binaries
go build -v -o output/linux_amd64/cnm/azure-vnet-plugin -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cnm/plugin/*.go
cnm/plugin/main.go:12:2: cannot find package "github.com/Azure/azure-container-networking/cnm/ipam" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/cnm/ipam (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/cnm/ipam (from $GOPATH)
cnm/plugin/main.go:13:2: cannot find package "github.com/Azure/azure-container-networking/cnm/network" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/cnm/network (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/cnm/network (from $GOPATH)
cnm/plugin/main.go:14:2: cannot find package "github.com/Azure/azure-container-networking/common" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/common (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/common (from $GOPATH)
cnm/plugin/main.go:15:2: cannot find package "github.com/Azure/azure-container-networking/log" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/log (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/log (from $GOPATH)
cnm/plugin/main.go:16:2: cannot find package "github.com/Azure/azure-container-networking/platform" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/platform (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/platform (from $GOPATH)
cnm/plugin/main.go:17:2: cannot find package "github.com/Azure/azure-container-networking/store" in any of:
        /usr/lib/golang/src/github.com/Azure/azure-container-networking/store (from $GOROOT)
        /home/user/go/src/github.com/Azure/azure-container-networking/store (from $GOPATH)
make: *** [output/linux_amd64/cnm/azure-vnet-plugin] Error 1
```
```
[user@localhost azure-container-networking]$ go get github.com/Azure/azure-container-networking
package github.com/Azure/azure-container-networking: no Go files in /home/user/go/src/github.com/Azure/azure-container-networking
```

```
[user@localhost ]$ cd /home/user/go/src/github.com/Azure/azure-container-networking
make all-binaries
go build -v -o output/linux_amd64/cnm/azure-vnet-plugin -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cnm/plugin/*.go
netlink/ip.go:9:2: cannot find package "golang.org/x/sys/unix" in any of:
        /usr/lib/golang/src/golang.org/x/sys/unix (from $GOROOT)
        /home/user/go/src/golang.org/x/sys/unix (from $GOPATH)
make: *** [output/linux_amd64/cnm/azure-vnet-plugin] Error 1
```

```
[user@localhost azure-container-networking]$ go get golang.org/x/sys/unix
make all-binaries
go build -v -o output/linux_amd64/cnm/azure-vnet-plugin -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cnm/plugin/*.go
github.com/Azure/azure-container-networking/platform
github.com/Azure/azure-container-networking/log
github.com/Azure/azure-container-networking/store
github.com/Azure/azure-container-networking/common
github.com/Azure/azure-container-networking/cnm
github.com/Azure/azure-container-networking/ipam
github.com/Azure/azure-container-networking/cnm/ipam
github.com/Azure/azure-container-networking/ebtables
github.com/Azure/azure-container-networking/netlink
github.com/Azure/azure-container-networking/network/policy
github.com/Azure/azure-container-networking/network
github.com/Azure/azure-container-networking/cnm/network
command-line-arguments
chmod 0755 output/linux_amd64/cnm/azure-vnet-plugin
cd output/linux_amd64/cnm && tar -czvf azure-vnet-cnm-linux-amd64-v1.0.5-1-g89555ac.tgz azure-vnet-plugin
azure-vnet-plugin
chown 1000:1000 output/linux_amd64/cnm/azure-vnet-cnm-linux-amd64-v1.0.5-1-g89555ac.tgz
go build -v -o output/linux_amd64/cni/azure-vnet -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cni/network/plugin/*.go
cni/plugin.go:16:2: cannot find package "github.com/containernetworking/cni/pkg/invoke" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/invoke (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/invoke (from $GOPATH)
cni/cni.go:7:2: cannot find package "github.com/containernetworking/cni/pkg/skel" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/skel (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/skel (from $GOPATH)
cni/internal.go:10:2: cannot find package "github.com/containernetworking/cni/pkg/types" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/types (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/types (from $GOPATH)
cni/plugin.go:19:2: cannot find package "github.com/containernetworking/cni/pkg/types/current" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/types/current (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/types/current (from $GOPATH)
cni/plugin.go:20:2: cannot find package "github.com/containernetworking/cni/pkg/version" in any of:
        /usr/lib/golang/src/github.com/containernetworking/cni/pkg/version (from $GOROOT)
        /home/user/go/src/github.com/containernetworking/cni/pkg/version (from $GOPATH)
make: *** [output/linux_amd64/cni/azure-vnet] Error 1
```

```
[user@localhost azure-container-networking]$ go get github.com/containernetworking/cni/pkg/invoke
[user@localhost  azure-container-networking]$ go get github.com/containernetworking/cni/pkg/skel
```

```
[user@localhost  azure-container-networking]$ make all-binaries
chmod 0755 output/linux_amd64/cnm/azure-vnet-plugin
cd output/linux_amd64/cnm && tar -czvf azure-vnet-cnm-linux-amd64-v1.0.5-1-g89555ac.tgz azure-vnet-plugin
azure-vnet-plugin
chown 1000:1000 output/linux_amd64/cnm/azure-vnet-cnm-linux-amd64-v1.0.5-1-g89555ac.tgz
go build -v -o output/linux_amd64/cni/azure-vnet -ldflags "-X main.version=v1.0.5-1-g89555ac -s -w" cni/network/plugin/*.go
github.com/Azure/azure-container-networking/platform
github.com/Azure/azure-container-networking/log
github.com/Azure/azure-container-networking/store
github.com/Azure/azure-container-networking/common
github.com/Azure/azure-container-networking/network/policy
github.com/Azure/azure-container-networking/cni
# github.com/Azure/azure-container-networking/cni
cni/plugin.go:112:39: not enough arguments in call to skel.PluginMainWithError
        have (func(*skel.CmdArgs) error, func(*skel.CmdArgs) error, version.PluginInfo)
        want (func(*skel.CmdArgs) error, func(*skel.CmdArgs) error, func(*skel.CmdArgs) error, version.PluginInfo)
github.com/Azure/azure-container-networking/ebtables
github.com/Azure/azure-container-networking/netlink
github.com/Azure/azure-container-networking/network
github.com/Azure/azure-container-networking/telemetry
make: *** [output/linux_amd64/cni/azure-vnet] Error 2
```

How to build this fucking shit?

### PR DESCRIPTION
**Pull request by ashvindeodhar**
CNI spec has added a new method : GET which allows the runtime / orchestrator to query the current state of the container networking configuration. This change adds the implementation for the GET method in azure-vnet plugin. The supported version changes to 0.4.0 to support GET API.
This PR fixes #148

### CODE REVIEW

File: network/endpoint.go [hunk 1]
```diff
@@ -144,6 +150,18 @@ func (ep *endpoint) getInfo() *EndpointInfo {
 		Id:          ep.Id,
 		IPAddresses: ep.IPAddresses,
 		Data:        make(map[string]interface{}),
+		MacAddress:  ep.MacAddress,
+		SandboxKey:  ep.SandboxKey,
+		IfIndex:     0, // Azure CNI supports only one interface
```
Comment (tamilmani1989) [inline network/endpoint.go]: why this value is hardcoded.? Whicn interface it points to?

Comment (tamilmani1989) [inline network/endpoint.go]: why this value is hardcoded.? Whicn interface it points to?

File: network/endpoint.go [hunk 2]
```diff
@@ -21,6 +21,8 @@ type endpoint struct {
 	MacAddress  net.HardwareAddr
 	IPAddresses []net.IPNet
 	Gateways    []net.IP
+	DNS         DNSInfo
```
Comment (tamilmani1989) [inline network/endpoint.go]: can we move Data to last field in structure..also plz group same datatypes together

File: cni/plugin.go [hunk 3]
```diff
@@ -177,3 +151,39 @@ func (plugin *Plugin) Error(err error) *cniTypes.Error {
...
+	if plugin.Store == nil {
+		var err error
+		plugin.Store, err = store.NewJsonFileStore(platform.CNIRuntimePath + plugin.Name + ".json")
+		if err != nil {
+			log.Printf("[cni] Failed to create store, err:%v.", err)
+			return err
+		}
+
+		// Acquire store lock.
+		err = plugin.Store.Lock(true)
```
Comment (tamilmani1989) [inline cni/plugin.go]: we started following this convention now.. in old code we dint change to this..but in new code we are following this
if err := function(); err != nil { 
}

File: cni/plugin.go [hunk 4]
```diff
@@ -177,3 +151,39 @@ func (plugin *Plugin) Error(err error) *cniTypes.Error {
...
+		if err != nil {
+			log.Printf("[cni] Timed out on locking store, err:%v.", err)
+			return err
+		}
+
+		config.Store = plugin.Store
+	}
```
Comment (tamilmani1989) [inline cni/plugin.go]: there should be a blank line before return

File: cni/plugin.go [hunk 5]
```diff
@@ -177,3 +151,39 @@ func (plugin *Plugin) Error(err error) *cniTypes.Error {
...
+// Uninitialize key-value store
+func (plugin *Plugin) UninitializeKeyValueStore() error {
+	if plugin.Store != nil {
+		err := plugin.Store.Unlock()
+		if err != nil {
+			log.Printf("[cni] Failed to unlock store, err:%v.", err)
+			return err
+		}
+	}
+	plugin.Store = nil
```
Comment (tamilmani1989) [inline cni/plugin.go]: blank line

File: cni/network/plugin/main.go [hunk 6]
```diff
@@ -83,11 +83,30 @@ func main() {
...
+			log.Printf("Failed to uninitialize key-value store of network plugin, err:%v.\n", err)
+		}
+
+		if recover() != nil {
+			os.Exit(1)
+		}
+	}()
+
+	err = netPlugin.Plugin.InitializeKeyValueStore(&config)
+
```
Comment (tamilmani1989) [inline cni/network/plugin/main.go]: remove blank and you can check for err in same line as i said below

File: network/endpoint.go (hunk 1 above)
Comment (ashvindeodhar) [reply network/endpoint.go]: hardcoded this to 0 with Sushant. Let me touch base with him again if this needs to be a declared const.

File: network/endpoint.go (hunk 2 above)
Comment (ashvindeodhar) [reply network/endpoint.go]: Done

File: cni/plugin.go (hunk 3 above)
Comment (ashvindeodhar) [reply cni/plugin.go]: Done

File: cni/plugin.go (hunk 4 above)
Comment (ashvindeodhar) [reply cni/plugin.go]: Done

File: cni/plugin.go (hunk 5 above)
Comment (ashvindeodhar) [reply cni/plugin.go]: Done

File: cni/network/plugin/main.go (hunk 6 above)
Comment (ashvindeodhar) [reply cni/network/plugin/main.go]: Done. Changed to this stype in other places as well

File: cni/plugin.go [hunk 7]
```diff
@@ -177,3 +151,39 @@ func (plugin *Plugin) Error(err error) *cniTypes.Error {
 func (plugin *Plugin) Errorf(format string, args ...interface{}) *cniTypes.Error {
 	return plugin.Error(fmt.Errorf(format, args...))
 }
+
+// Initialize key-value store
+func (plugin *Plugin) InitializeKeyValueStore(config *common.PluginConfig) error {
```
Comment (saiyan86) [inline cni/plugin.go]: Why do we need this? We are already saving the state of networkManager. What happens if we get a GET call before any ADD happens? Even in this case I don't think we need to initialize json store here, as we can print an empty result.

File: cni/plugin.go [hunk 8]
```diff
@@ -177,3 +151,39 @@ func (plugin *Plugin) Error(err error) *cniTypes.Error {
...
+			return err
+		}
+
+		config.Store = plugin.Store
+	}
+	return nil
+}
+
+// Uninitialize key-value store
+func (plugin *Plugin) UninitializeKeyValueStore() error {
```
Comment (saiyan86) [inline cni/plugin.go]: Same here.

File: cni/plugin.go (hunk 7 above)
Comment (sharmasushant) [reply cni/plugin.go]: This initialize is called from the main. We can either check for the operation type in main and then avoid initialization, or we can always initialize and do the operation to keep logic simple. i'm fine with either.

**Pull request merged**